presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java

```diff
@@ -284,36 +283,88 @@ private void checkEncodeInputValidity(List<Integer> input, int maximumBits)
+        List<BoundedZValueRange> boundedValueRanges = fillUnspecifiedRangesWithDefaults(valueRanges);
 
-        findAddressesRecursively(maxBitLength - 1, 0L, totalBitLength, fillUnspecifiedRangesWithDefaults(valueRanges), addressRanges);
+        int[] lengths = new int[valueRanges.size()];
+        for (int index = 0; index < valueRanges.size(); index++) {
+            lengths[index] = valueRanges.get(index).length;
+        }
+
+        int[] counters = new int[valueRanges.size()];
+        Arrays.fill(counters, 0);
+
+        List<ZAddressRange<Long>> addressRanges = new ArrayList<>();
+        permuteValueRanges(counters, lengths, 0, boundedValueRanges, addressRanges);
 
         return combineAddressRanges(addressRanges);
     }
 
-    private List<ZValueRange> fillUnspecifiedRangesWithDefaults(List<ZValueRange> valueRanges)
+    private List<BoundedZValueRange> fillUnspecifiedRangesWithDefaults(List<ZValueRange> valueRanges)
     {
-        List<ZValueRange> nonNullValues = new ArrayList<>(valueRanges.size());
+        List<BoundedZValueRange> nonNullValues = new ArrayList<>(valueRanges.size());
         for (int index = 0; index < valueRanges.size(); index++) {
             ZValueRange range = valueRanges.get(index);
 
-            Integer minimumValue = range.getMinimumValue();
-            Integer maximumValue = range.getMaximumValue();
-
-            if (minimumValue == null) {
-                minimumValue = 0;
+            List<Integer> minimumValues = new ArrayList<>();
+            for (Integer minimumValue : range.getMinimumValues()) {
+                if (minimumValue == null) {
+                    minimumValue = 0;
+                }
+                minimumValues.add(minimumValue);
             }
-            if (maximumValue == null) {
-                maximumValue = (1 << encodingBits.get(index)) - 1;
+
+            List<Integer> maximumValues = new ArrayList<>();
+            for (Integer maximumValue : range.getMaximumValues()) {
+                if (maximumValue == null) {
+                    maximumValue = (1 << encodingBits.get(index)) - 1;
+                }
+                maximumValues.add(maximumValue);
             }
 
-            nonNullValues.add(new ZValueRange(Optional.of(minimumValue), Optional.of(maximumValue)));
+            nonNullValues.add(new BoundedZValueRange(minimumValues, maximumValues));
         }
         return nonNullValues;
     }
 
+    private void permuteValueRanges(int[] counters, int[] lengths, int level, List<BoundedZValueRange> valueRanges, List<ZAddressRange<Long>> addressRanges)
+    {
+        if (level == counters.length) {
+            findAddressesRecursively(maxBitLength - 1, 0L, totalBitLength, valueRanges, addressRanges, counters);
+        }
+        else {
+            for (counters[level] = 0; counters[level] < lengths[level]; counters[level]++) {
+                permuteValueRanges(counters, lengths, level + 1, valueRanges, addressRanges);
+            }
+        }
```

Comment:
Let us change it a little bit in format as follows:

if (....) {
....
return;
}

for (....) {
}